libpromises/mod_custom.c

```diff
@@ -0,0 +1,651 @@
+        policy, pp->parent_section->promise_type);
+}
+
+static Body *FindCustomPromiseType(const Promise *promise)
+{
+    assert(promise != NULL);
+
+    const char *const promise_type = promise->parent_section->promise_type;
+    const Policy *const policy =
+        promise->parent_section->parent_bundle->parent_policy;
+    Seq *custom_promise_types = policy->custom_promise_types;
+    const size_t length = SeqLength(custom_promise_types);
+    for (int i = 0; i < length; ++i)
+    {
+        Body *current = SeqAt(custom_promise_types, i);
+        if (StringEqual(current->name, promise_type))
+        {
+            return current;
+        }
+    }
+    return NULL;
+}
+
+static bool GetInterpreterAndPath(
+    Body *promise_block, char **interpreter_out, char **path_out)
+{
+    assert(promise_block != NULL);
+    assert(interpreter_out != NULL);
+    assert(path_out != NULL);
+
+    char *interpreter = NULL;
+    char *path = NULL;
+
+    const char *promise_type = promise_block->name;
+    Seq *promise_block_attributes = promise_block->conlist;
+    const size_t length = SeqLength(promise_block_attributes);
+
+    for (int i = 0; i < length; ++i)
+    {
+        Constraint *attribute = SeqAt(promise_block_attributes, i);
+        const char *name = attribute->lval;
+        char *value = RvalToString(attribute->rval);
+
+        if (StringEqual("interpreter", name))
+        {
+            free(interpreter);
+            interpreter = value;
+        }
+        else if (StringEqual("path", name))
+        {
+            free(path);
+            path = value;
+        }
+        else
+        {
+            debug_abort_if_reached();
+            free(value);
+        }
+    }
+
+    if ((interpreter == NULL) || (path == NULL))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Custom promise type '%s' missing interpreter or path",
+            promise_type);
+        free(interpreter);
+        free(path);
+        return false;
+    }
+
+    *interpreter_out = interpreter;
+    *path_out = path;
+    return true;
+}
+
+static void PromiseModule_LogJson(JsonElement *object)
+{
+    const char *level_string = JsonObjectGetAsString(object, "level");
+    const char *message = JsonObjectGetAsString(object, "message");
+
+    assert(level_string != NULL && message != NULL);
+    const LogLevel level = LogLevelFromString(level_string);
+    assert(level != LOG_LEVEL_NOTHING);
+    Log(level, "%s", message);
+}
+
+static JsonElement *PromiseModule_Receive(PromiseModule *module)
+{
+    assert(module != NULL);
+
+    bool line_based = !(module->json);
+
+    char *line = NULL;
+    size_t size = 0;
+    ssize_t read;
+    bool empty_line = false;
+    JsonElement *log_array = JsonArrayCreate(10);
+    JsonElement *response = NULL;
+
+    if (line_based)
+    {
+        response = JsonObjectCreate(10);
+    }
+
+    while (!empty_line && ((read = getline(&line, &size, module->output)) > 0))
+    {
+        assert(read > 0);
```

Comment:
`read` hides the `read()` syscall/function.

Reply:
Fixed.